scipy/optimize/_gensa.py

```diff
@@ -0,0 +1,669 @@
+        """
+        # It it is the first part of the markov chain
+        # Changing all components at the same time
+        if step < self._x.size:
+            visits = np.array([self._visita() for _ in range(self._x.size)])
+            visits[visits > 1.e8] = 1.e8 * self._random_state.random_sample()
+            visits[visits < -1e8] = -1.e8 * self._random_state.random_sample()
+            self._x = visits + self._xbackup
+            a = self._x - self._lower
+            b = np.fmod(a, self._xrange) + self._xrange
+            self._x = np.fmod(b, self._xrange) + self._lower
+            self._x[np.fabs(self._x - self._lower) < 1.e-10] += 1.e-10
+        else:
+            # Second part of the markov chain
+            # Now change only one component at a time
+            visit = self._visita()
+            if visit > 1.e8:
+                visit = 1.e8 * self._random_state.random_sample()
+            elif visit < -1e8:
+                visit = -1.e8 * self._random_state.random_sample()
+            index = step - self._x.size
+            self._x[index] = visit + self._xbackup[index]
+            a = self._x[index] - self._lower[index]
+            b = np.fmod(
+                a, self._xrange[index]) + self._xrange[index]
+            self._x[index] = np.fmod(
+                b, self._xrange[index]) + self._lower[index]
+            if np.fabs(self._x[index] - self._lower[
+                    index]) < 1.e-10:
+                self._x[index] += 1.e-10
+
+    def _run_markov_chain(self, step):
+        for j in range(self.markov_length):
+            if j == 0:
+                self._emin_unchanged = True
+            if step == 0 and j == 0:
+                self._emin_unchanged = False
+            # Keeping old location before visiting new place
+            self._xbackup = np.array(self._x)
+            self._visiting(j)
+            self._energy(self._x)
+            if self._etot < self._etot0:
+                # We get a better energy value
+                self._etot0 = np.array(self._etot)
+                if self._etot < self._emin:
+                    self._emin = np.array(self._etot)
+                    self._xmin = np.array(self._x)
+                    self._emin_unchanged = False
+                    self._index_no_emin_update = 0
+            else:
+                # We do not have improvement but do we accept location
+                r = self._random_state.random_sample()
+                pqa1 = (self.qa - 1.0) * (
+                    self._etot - self._etot0) / self._temperature_qa + 1.0
+                if pqa1 < 0.0:
+                    pqa = 0.0
+                else:
+                    pqa = np.exp(np.log(pqa1) / (1.0 - self.qa))
+                if r > pqa:
+                    # We reject the new visiting location
+                    self._x = self._xbackup
+                else:
+                    # The new visiting location is accepted
+                    self._etot0 = self._etot
```

Comment:
why do you write it like this here when you did `self._etot0 = np.array(self._etot)` above?

In general I think it would be better structured as (pseudo code)

```
    If _new_location_is_accepted(xnew, xold, enew, eold):
        // update x, energy, and xmin, emin if appropriate
    else:
        self._x = self._xbackup
```